include/RAJA/index/ListSegment.hpp

```diff
@@ -93,202 +96,206 @@ class ListSegment : public BaseSegment
   ///
-  ListSegment(const Index_type* indx,
-              Index_type len,
-              IndexOwnership indx_own = Owned);
+  TypedListSegment(const value_type* values,
+                   Index_type length,
+                   IndexOwnership owned = Owned)
+  {
+    initIndexData(values, length, owned);
+  }
 
   ///
   /// Construct list segment from arbitrary object holding
   /// indices using a deep copy of given data.
   ///
   /// The object must provide methods: begin(), end(), size().
   ///
-  template <typename T>
-  explicit ListSegment(const T& indx);
+  template <typename Container>
+  explicit TypedListSegment(const Container& container)
+      : m_data(0), m_size(container.size()), m_owned(Unowned)
+  {
+    if (container.size() > 0) {
+#if defined(RAJA_ENABLE_CUDA)
+      cudaErrchk(cudaMallocManaged((void**)&m_data,
+                                   m_size * sizeof(T),
+                                   cudaMemAttachGlobal));
+      using ContainerType =
+          typename std::remove_const<typename std::remove_reference<decltype(
+              *container.begin())>::type>::type;
+      if (!std::is_pointer<ContainerType>::value
+          && sizeof(ContainerType) == sizeof(value_type)) {
+        // this will bitwise copy signed or unsigned integral types
+        cudaErrchk(cudaMemcpy(m_data,
+                              std::addressof(container[0]),
+                              m_size * sizeof(value_type),
+                              cudaMemcpyDefault));
+      } else {
+        cudaErrchk(cudaDeviceSynchronize());
+        auto self = m_data;
+        auto other = container.begin();
+        auto end = container.end();
+        while (other != end) {
+          *self = *other;
+          ++self;
+          ++other;
+        }
+      }
+#else
+      m_data = new value_type[container.size()];
+      auto self = m_data;
+      auto other = container.begin();
+      auto end = container.end();
+      while (other != end) {
+        *self = *other;
+        ++self;
+        ++other;
+      }
+#endif
+      m_owned = Owned;
+    }
+  }
 
   ///
   /// Copy-constructor for list segment.
   ///
-  ListSegment(const ListSegment& other);
+  TypedListSegment(const TypedListSegment& other)
+  {
+    initIndexData(other.m_data, other.m_size, other.m_owned);
+  }
 
   ///
   /// Copy-assignment for list segment.
   ///
-  ListSegment& operator=(const ListSegment& rhs);
+  TypedListSegment& operator=(const TypedListSegment& rhs)
+  {
+    if (&rhs != this) {
+      TypedListSegment copy(rhs);
+      this->swap(copy);
+    }
+    return *this;
+  }
 
   ///
   /// Move-constructor for list segment.
   ///
-  ListSegment(ListSegment&& other)
-      : BaseSegment(std::move(other)),
-        m_indx(other.m_indx),
-        m_len(other.m_len),
-        m_indx_own(other.m_indx_own)
-  {
-    other.m_indx = nullptr;
-  }
+  TypedListSegment(TypedListSegment&&) = default;
```

Comment:
Update: yes you are correct here. I'll fix this. I rolled my own smoke test to verify